go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -75,30 +75,48 @@ type serviceConfig struct {
-// This function is currently used only in LGW mode and when ETP=local for the service
-// cookie is the hash value of openflow cookie, svcPort is the nodePort of the service, flowProtocol is tcp/udp/sctp in v4 or v6 modes,
+// createFlowsForLocalHostNetEp adds the necessary openflows for a service when its backed by local-host-networked endpoints
+// This function is currently used only when ETP=local is set for the service
+// cookie is the hash value of openflow cookie, svcPort is the kapi.ServicePort object, flowProtocol is tcp/udp/sctp in v4 or v6 modes,
 // externalIPOrLBIngressIP is either externalIP or LB.status.ingress, nwDst&nwSrc are constants used to match on the destinations based on the flowprotocol
-func (npw *nodePortWatcher) createFlowsForNonLocalHostNetEp(cookie string, svcPort int32, flowProtocol string, externalIPOrLBIngressIP, nwDst, nwSrc string) []string {
-	var nodeportFlowsResult []string
+func (npw *nodePortWatcher) createFlowsForLocalHostNetEp(cookie string, svcPort *kapi.ServicePort, flowProtocol string, externalIPOrLBIngressIP, nwDst, nwSrc string) []string {
+	var nodeportFlows []string
 	if len(externalIPOrLBIngressIP) != 0 {
 		// if externalIPOrLBIngressIP is set then this service is of type LB/EIP
-		nodeportFlowsResult = append(nodeportFlowsResult,
-			fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,table=6)",
-				cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort, HostNodePortCTZone),
-			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, %s=%s, tp_src=%d, actions=ct(zone=%d,table=7)",
-				cookie, flowProtocol, nwSrc, externalIPOrLBIngressIP, svcPort, HostNodePortCTZone))
+		// table 0, This rule matches on all traffic with dst ip == LoadbalancerIP / externalIP, DNAT's it to the correct NodeIP
+		if strings.Contains(flowProtocol, "6") {
+			// If ipv6 make sure to choose the ipv6 node address for rule
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=[%s]:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv6, svcPort.TargetPort.String()))
+		} else {
+			nodeportFlows = append(nodeportFlows,
+				fmt.Sprintf("cookie=%s, priority=110, in_port=%s, %s, %s=%s, tp_dst=%d, actions=ct(commit,zone=%d,nat(dst=%s:%s),table=6)",
+					cookie, npw.ofportPhys, flowProtocol, nwDst, externalIPOrLBIngressIP, svcPort.Port, HostNodePortCTZone, npw.gatewayIPv4, svcPort.TargetPort.String()))
+		}
+		nodeportFlows = append(nodeportFlows,
+			// table 0, Matches on return traffic, i.e traffic coming from the host networked pod's port, and unDNATs
+			fmt.Sprintf("cookie=%s, priority=110, in_port=LOCAL, %s, tp_src=%s, actions=ct(commit,zone=%d nat,table=7)",
+				cookie, flowProtocol, svcPort.TargetPort.String(), HostNodePortCTZone))
 	} else {
```

Comment:
wait, cant a service have external IP, LB ingress IP, and also be nodeport? why is this else here? @astoycos ?

Reply:
oh I think this is the semantic of the function, if externalIPOrLBIngressIP is empty, then assume its nodeport. I think we should change this function and re-examine this code. This now seems overly complicated and not very clear.

Reply:
@trozet maybe I should not modularize this :D I think that will be more clear, wdyt?

Reply:
done.